Use _Py_DebugOffsets instead of hardcoded offsets when possible

godlygeek opened the pull request:
Python 3.13 adds a new `_Py_DebugOffsets` structure to `_PyRuntimeState` which tells us more or less everything that we need to know in order to decode the interpreter's data structures. Look for this structure and, if we find it, use the offsets contained in it rather than the offsets compiled into pystack at build time.

This allows us to decode stacks for interpreters with different build time sizes of different structures, including those running in a `python3.13t` free-threaded interpreter.

pablogsal commented on src/pystack/_pystack/version.h:
```diff
@@ -138,13 +125,15 @@ struct py_thread_v
     FieldOffset<remote_addr_t> o_finalizing;
     FieldOffset<remote_addr_t> o_interp_head;
-    FieldOffset<GCRuntimeState> o_gc;
+    FieldOffset<char> o_gc;
```
Hummm why `char`? Isn't this a pointer?

godlygeek replied:
Yeah, but it's an inline structure whose layout we don't know at compile time, and so there's not a useful version-independent type we can use to refer to it by here. `char` is a hack to deal with the fact that we can't actually dereference this field into a meaningful `GCRuntimeState` directly, and instead need to take a pointer to it and create a `Structure` from that pointer.

I'll add a comment explaining this.

pablogsal commented on src/pystack/_pystack/structure.h:
```diff
@@ -0,0 +1,85 @@
+Structure<OffsetsStruct>::getField(FieldPointer OffsetsStruct::*field)
+{
+    copyFromRemote();
+    offset_t offset = (d_manager->offsets().get<OffsetsStruct>().*field).offset;
```
Should we add an assert to bound check against `d_buf`?

pablogsal commented on tests/integration/test_gil.py:
```diff
@@ -14,6 +17,15 @@
+@pytest.fixture(autouse=True)
+def enable_gil_if_free_threading(python, monkeypatch):
+    _, python_executable = python
+    proc = subprocess.run([python_executable, "-Xgil=1", "-cpass"], capture_output=True)
```
This will always succeed as an invalid `-X` option doesn't raise or warn. I am missing something here?

godlygeek replied:
An unrecognized `-X` option doesn't raise or warn, but `python3.13` recognizes `-Xgil` and fails when the build is not free-threading:
```
$ python3.13 -Xgil=1
Fatal Python error: config_read_gil: PYTHON_GIL / -X gil are not supported by this build
Python runtime state: preinitialized
```

godlygeek replied:
I've opened an upstream bug report for this annoying behavior, https://github.com/python/cpython/issues/123275

pablogsal commented on tests/integration/test_local_variables.py:
```diff
@@ -573,31 +573,33 @@ def test_trashed_locals(generate_threads, python, tmpdir):
         ("ob_item1", ctypes.c_void_p),
     ]
 
+def ob_type_field(obj):
```
Is this something we can avoid with the new offsets we added to 3.13?

godlygeek replied:
It's technically possible, but would require much more complexity. We'd need a way for pystack to attach to a running process and dump the offsets rather than the stack, and we'd need to then take those offsets and pass them to program being run by the test.

We've got options if this turns out to be a maintenance burden, but for now, I'd rather keep this as simple as we can.

pablogsal reviewed (approved):
LGTM modulo the things that we discussed offline

pablogsal commented on src/pystack/process.py:
```diff
@@ -20,7 +20,8 @@
 
 BSS_VERSION_REGEXP = re.compile(
-    rb"((2|3)\.(\d+)\.(\d{1,2}))((a|b|c|rc)\d{1,2})?\+? (\(.{1,64}\))"
+    rb"((2|3)\.(\d+)\.(\d{1,2}))((a|b|c|rc)\d{1,2})?\+?"
+    rb"(?: experimental free-threading build)? (\(.{1,64}\))"
```
```suggestion
    rb"""
    (                   # Capturing group for the main version
        (2|3)           # Major version: 2 or 3
        \.              # Literal dot
        (\d+)           # Minor version: one or more digits
        \.              # Literal dot
        (\d{1,2})       # Patch version: 1 or 2 digits
    )
    (                   # Capturing group for pre-release identifier (optional)
        (a|b|c|rc)      # Pre-release type: alpha, beta, candidate, or release candidate
        \d{1,2}         # Pre-release version: 1 or 2 digits
    )?                  # The entire pre-release group is optional
    \+?                 # Optional plus sign for build metadata
    (?:                 # Non-capturing group for experimental build info (optional)
        \s              # Space
        experimental\ free-threading\ build
    )?
    \s                  # Space
    (                   # Capturing group for additional build info
        \(              # Opening parenthesis
        .{1,64}         # 1 to 64 characters of any type (except newline)
        \)              # Closing parenthesis
    )
    """,
    re.VERBOSE | re.BINARY
```

pablogsal replied:
Matt doesn't like my suggestion :(

pablogsal replied:
maybe add some examples:

```
# Sample strings that match the regex:
#
# b"3.8.10 (default, May 26 2023, 14:05:08)"
# b"2.7.18rc1 (v2.7.18rc1:8d21aa21f2, Apr 20 2020, 13:19:08)"
# b"3.9.5+ experimental free-threading build (Python)"
# b"3.10.0b2 (main, Jun 22 2021, 12:46:30)"
# b"2.7.15+ (default, Mar  5 2023, 16:38:35)"
```

godlygeek replied:
Comments shouldn't explain _what_ the code is doing, they should explain _why_ it's doing those things 😄 

To that end, the example strings are much more instructive than the plain English explanation of the regex syntax. I've taken 3 of them that seem to cover all the interesting cases.

godlygeek commented:
@pablogsal I've addressed all of your suggestions. I've also rebased this on main, and fixed conflicts that arose due to the fix for handling shim frames. Take another look when you get a chance, please.
